### PR TITLE
ci: CI docker TAV tests are unnecessarily setting up all services for some modules

### LINKED ISSUE #2061: the Windows-Node.js-12 jenkins test is failing
From https://apm-ci.elastic.co/job/apm-agent-nodejs/job/apm-agent-nodejs-mbp/job/master/

Screen Shot 2021-04-26 at 1 46 10 PM

it last passed in build 864 on Apr 21st. The next two builds aborted in the "Windows-Node.js-12" step after ~3h (timeout?) -- I'm not sure what those are about. The next two (builds 867 and 868) failed in that same step with this in the log:

```
...
[2021-04-23T21:19:49.079Z] Getting latest nodejs-lts version for 12 ...
[2021-04-23T21:19:55.681Z] Installing nodejs-lts version:  ...
[2021-04-23T21:19:55.681Z] Chocolatey v0.10.15
[2021-04-23T21:19:55.944Z] Install Command
[2021-04-23T21:19:55.944Z] 
...
[2021-04-23T21:19:55.950Z] Package name is required. Please pass at least one package name to install.
script returned exit code 1
...
```

For comparison the previous *working* equivalent step:

Screen Shot 2021-04-26 at 1 49 21 PM

This is running this function from apm-pipeline-library: https://github.com/elastic/apm-pipeline-library/blob/master/resources/scripts/install-with-choco.ps1
Specifically I believe it is getting and empty string from this command:
https://github.com/elastic/apm-pipeline-library/blob/64546ee45b1c64b11d6138053401062d51c7e36f/resources/scripts/install-with-choco.ps1#L20

Chocolatey no longer lists node v12 versions in the "nodejs-lts" package set (https://community.chocolatey.org/packages?q=nodejs-lts):

```
PS C:\WINDOWS\system32> choco list nodejs-lts --exact --by-id-only --all
Chocolatey v0.10.15
nodejs-lts 14.16.1 [Approved]
1 packages found.
```

The `$pattern` we currently have is `12`, we I gather we are getting an empty list here.
The result is calling `choco install ...` without a package name (https://github.com/elastic/apm-pipeline-library/blob/64546ee45b1c64b11d6138053401062d51c7e36f/resources/scripts/install-with-choco.ps1#L52-L53), which fails.

### PR DESCRIPTION
**Pull request by trentm**
Invert logic of the selection of an appropriate docker-compose config
for *TAV* testing in docker, to default to *not setting up any extra
services* for a test run, unless specifically required for that TAV
module. Before this, TAV tests for `@hapi/hapi`, `@koa/router`, and
`@elastic/elasticsearch` would setup *all* services, which is overkill.

Also document .ci/scripts/test.sh because it subtlely does a few
slightly different things. Also rename the TAV_MODULE var name to be
more accurate.

Also bumps to using node 14 for the windows build which fixes #2061

* * *

Looking at the log for https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-nodejs%2Fapm-agent-nodejs-mbp/detail/master/864/pipeline/1037
which is step "TAV Test / Node.js-10-@hapi/hapi" of the TAV tests:

```
[2021-04-22T06:05:24.749Z] + .ci/scripts/test.sh 10 @hapi/hapi false
[2021-04-22T06:05:24.749Z] + DOCKER_FOLDER=.ci/docker
[2021-04-22T06:05:24.749Z] + NODE_VERSION=10
[2021-04-22T06:05:24.749Z] + TAV_VERSIONS=@hapi/hapi
[2021-04-22T06:05:24.749Z] + IS_EDGE=false
...
[2021-04-22T06:05:24.749Z] + docker-compose --no-ansi --log-level ERROR -f .ci/docker/docker-compose-all.yml build
...
```

That is using the docker compose file to setup *all* services for testing, which is overkill. That is a failing of this block that selects an appropriate docker compose config for the test: https://github.com/elastic/apm-agent-nodejs/blob/v3.14.0/.ci/scripts/test.sh#L9-L40

If we grep the full [consoleText of build 864](https://apm-ci.elastic.co/job/apm-agent-nodejs/job/apm-agent-nodejs-mbp/job/master/864/consoleText) for 'docker-compose ... build' usage (full grep output here: https://gist.github.com/trentm/dd290655f594c6d162453c0ffb15d9a2), we see the following that are using `docker-compose-all.yml` unnecessarily:

```
% rg '.ci/scripts/test.sh|docker-compose .*build$' consoleText-build-864.log
...
275569:[2021-04-22T06:03:03.821Z] + .ci/scripts/test.sh 14 @elastic/elasticsearch false
275585:[2021-04-22T06:03:03.821Z] + docker-compose --no-ansi --log-level ERROR -f .ci/docker/docker-compose-all.yml build
275645:[2021-04-22T06:03:07.749Z] + .ci/scripts/test.sh 14 @hapi/hapi false
275661:[2021-04-22T06:03:07.749Z] + docker-compose --no-ansi --log-level ERROR -f .ci/docker/docker-compose-all.yml build
275673:[2021-04-22T06:03:09.194Z] + .ci/scripts/test.sh 14 @koa/router false
275689:[2021-04-22T06:03:09.194Z] + docker-compose --no-ansi --log-level ERROR -f .ci/docker/docker-compose-all.yml build
279407:[2021-04-22T06:03:56.116Z] + .ci/scripts/test.sh 13 @hapi/hapi false
279423:[2021-04-22T06:03:56.117Z] + docker-compose --no-ansi --log-level ERROR -f .ci/docker/docker-compose-all.yml build
279639:[2021-04-22T06:03:57.470Z] + .ci/scripts/test.sh 13 @koa/router false
279655:[2021-04-22T06:03:57.470Z] + docker-compose --no-ansi --log-level ERROR -f .ci/docker/docker-compose-all.yml build
279879:[2021-04-22T06:04:00.097Z] + .ci/scripts/test.sh 13 @elastic/elasticsearch false
279895:[2021-04-22T06:04:00.097Z] + docker-compose --no-ansi --log-level ERROR -f .ci/docker/docker-compose-all.yml build
286755:[2021-04-22T06:04:46.195Z] + .ci/scripts/test.sh 12 @elastic/elasticsearch false
286771:[2021-04-22T06:04:46.196Z] + docker-compose --no-ansi --log-level ERROR -f .ci/docker/docker-compose-all.yml build
... etc.
```

In other words, what happened is that support for `@hapi/hapi`, `@koa/router`, and `@elastic/elasticsearch` was added, this block in ".ci/scripts/tests.sh" was not updated, and the fallback behaviour of setting up *all* services for a test "worked". So it wasn't noticed that this logic should be updated for a new module instrumentation.

I'd like to invert the logic to setup *no* extra services for a `TAV_VERSIONS` by default. That means that the TAV tests for a newly added module instrumentation (and a matching new ".tav.yml" block) *that requires a service for testing* will **break** until .ci/scripts/test.sh is updated.

### CODE REVIEW

File: .ci/Jenkinsfile [hunk 1]
```diff
@@ -90,8 +90,8 @@ pipeline {
                 // parallelTasks["Windows-Node.js-${version}"] = generateStepForWindows(version: version)
               }
 
-              // Only 12 for the time being
-              parallelTasks["Windows-Node.js-12"] = generateStepForWindows(version: '12')
+              // Only 14 for the time being
+              parallelTasks["Windows-Node.js-14"] = generateStepForWindows(version: '14')
```
Comment (trentm) [inline .ci/Jenkinsfile:95]: This is a workaround/fix for https://github.com/elastic/apm-agent-nodejs/issues/2061

**Comment by trentm** [comment]
Test run https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-nodejs%2Fapm-agent-nodejs-mbp/detail/PR-2060/6/pipeline/949 shows that TAV tests work and the "@elastic/elasticsearch" tav test is now only setting up the elasticsearch service, as intended.

**Review by astorm — APPROVED** [review]
Approving, does what it says and I don't see any lingering `TAV_VERSIONS` in the repo so everything should keep running.

One thing that gives me slight pause, if I'm reading this correctly, is that when we add a new module to the TAV suite that it will also require us making a new entry to this switch statement if a service is required, and with TAV tests not running on every PR this requirement might not be obvious.   If there's a way the main, non-TAV suite could fail fast in this scenario and make it obvious that a new docker-compose-[module].yml is needed that help us-or-future-folks from stubbing their toe on the CI.

**Comment by trentm** [comment]
>  is that when we add a new module to the TAV suite that it will also require us making a new entry to this switch statement if a service is required, and with TAV tests not running on every PR this requirement might not be obvious

Two things:

1. I think as authors and reviewers that if we add a *new instrumented module* we should definitely make sure the TAV tests for it are run on the PR.
2. I *think* any update to files under "{test,lib}/instrumentation/modules" will result in TAV tests being run. At least that is how I read https://github.com/elastic/apm-agent-nodejs/blob/4019318a3c3f9386daf5449b944196a4dd056a7f/.ci/Jenkinsfile#L51-L55 (to set `TAV_UPDATED`) and https://github.com/elastic/apm-agent-nodejs/blob/4019318a3c3f9386daf5449b944196a4dd056a7f/.ci/Jenkinsfile#L123 to use `TAV_UPDATED` to decide whether to run the tav tests.
3. (https://www.youtube.com/watch?v=sAn7baRbhx4) A lofty/dreamy goal of mine with test suite updates (node-tap, and features to `tav`) might be to run a subset of tav tests for all PRs. That's a stretch however. :)
